File: advent2022/day11.py

from collections import namedtuple, deque
from itertools import takewhile
import operator
from typing import List, Tuple, TypeVar, Self, TextIO


OPS = {
    0: operator.add,
    1: operator.sub,
    2: operator.mul,
    3: operator.floordiv,
    4: lambda x, y: int((x % y) == 0),
    5: operator.pow,
}


class Op:
    PLUS = 0
    SUBTRACT = 1
    MULTIPLY = 2
    DIVIDE = 3
    DIVISIBLE_BY = 4
    POWER = 5

    @staticmethod
    def exec(op: int, worry: int, value: int) -> int:
        f = OPS[op]
        return f(worry, value)

    @staticmethod
    def parse(input_str: str) -> Tuple[int, int]:
        parts = input_str.split()
        op = None

        if parts[-2] == '+':
            op = (Op.PLUS, int(parts[-1]))
        elif parts[-2] == '*':
            if parts[-1] == 'old':
                op = (Op.POWER, 2)
            else:
                oip = (Op.MULTIPLY, int(parts[-1]))
        elif parts[-3] == 'divisible':
            op = (Op.DIVISIBLE_BY, int(parts[-1]))

        return op


class Monkey:

    def __init__(
        self, 
        n: int, 
        items: List[int], 
        op: Tuple[int, int], 
        test: Tuple[int, int],
        action: Tuple[int, int],
    ):
        self.n = n
        self.items = deque(items)
        self.op = op
        self.test = test
        self.action = action
        self.inspected = 0

    @staticmethod
    def parse(lines: List[str]) -> Self:
        n = None
        items = deque()
        op = None
        test = None
        action_true = None
        action_false = None

        for line in lines:
            line = line.strip()
            if line.startswith('Monkey '):
                n = int(line[7:-1])
            elif line.startswith('Starting items: '):
                items = deque(int(item.strip()) for item in line[16:].split(','))
            elif line.startswith('Operation: '):
                op = Op.parse(line[11:])
            elif line.startswith('Test: '):
                test = Op.parse(line[6:])
            elif line.startswith('If true: '):
                action_true = int(line.split()[-1])
            elif line.startswith('If false: '):
                action_false = int(line.split()[-1])

        return Monkey(n, items, op, test, (action_true, action_false))

    def inspect(self, worry: int) -> Tuple[int, int]:
        (op, value) = self.op
        worry = Op.exec(op, worry, value)
        worry = worry // 3

        (test, value) = self.test
        (action_true, action_false) = self.action
        if Op.exec(test, worry, value):
            target = action_true
        else:
            target = action_false

        self.inspected += 1
        return (target, worry)
        
    def inspect_all(self) -> List[Tuple[int, int]]:
        output = []

        while self.items:
            current = self.items.popleft()
            pair = self.inspect(current)
            output.append(pair)

        return output


def run_a(monkey_data: TextIO) -> int:
    lines = iter(monkey_data)
    monkeys = []
    while True:
        block = list(takewhile(lambda l: l.strip(), lines))
        if not block:
            break
        monkeys.append(Monkey.parse(block))

    for n in range(20):
        for monkey in monkeys:
            for target, worry in monkey.inspect_all():
                monkeys[target].items.append(worry)

    monkeys.sort(reverse=True, key=operator.attrgetter('inspected'))
    a, b = monkeys[:2]

    return a.inspected * b.inspected


def main():
    with open('./src/main/resources/data/day11.txt') as f:
        print(run_a(f))
        # 120384


if __name__ == '__main__':
    main()

File: src/main/kotlin/com/ericrochester/advent2022/Day11.kt
package com.ericrochester.advent2022

import kotlin.math.roundToInt
import kotlin.math.pow

class Day11: DayRuns<Int,Int> {
    val maxRoundCount = 20

    override fun runA(inputData: String): Int {
        val monkeys = paragraphs(inputData.lines()).map { Monkey.parse(it) }
        repeat(20) {
            for (monkey in monkeys) {
                for (throws in monkey.move()) {
                    monkeys[throws.target].catchItem(throws.worryAmount)
                }
            }
            dumpRound(it, monkeys)
        }
        dumpItemsSeen(monkeys)

        val busiest = monkeys.sortedByDescending { it.itemsSeen }.take(2)
        return busiest[0].itemsSeen * busiest[1].itemsSeen
    }

    private fun dumpRound(i: Int, monkeys: List<Monkey>) {
        println("Round $i")
        for (monkey in monkeys) {
            println("Monkey ${monkey.n}: ${monkey.items.joinToString(", ")}")
        }
        println()
    }

    private fun dumpItemsSeen(monkeys: List<Monkey>) {
        for (monkey in monkeys) {
            println("Monkey ${monkey.n} inspected items ${monkey.itemsSeen} times.")
        }
    }

    override fun runB(inputData: String): Int {
        TODO("Not yet implemented")
    }

}

data class Monkey(
    val n: Int,
    val items: MutableList<Int>,
    val worryOp: OpPair,
    val testOp: OpPair,
    val testThrows: ThrowTo
) {
    var itemsSeen: Int = 0

    fun move(): List<ThrowInfo> {
        val throws = mutableListOf<ThrowInfo>()

        while (items.size > 0) {
            var current = items.removeFirst()
            itemsSeen += 1
            current = worryOp.exec(current)
            current = current.floorDiv(3)
            val target = if (testOp.exec(current) == 0) {
                testThrows.ifTrue
            } else {
                testThrows.ifFalse
            }
            throws.add(ThrowInfo(target, current))
        }

        println("### ${n}. $throws")
        return throws
    }

    fun catchItem(item: Int) {
        items.add(item)
    }

    companion object {
        fun parse(lines: List<String>): Monkey {
            var n: Int = -1
            var items: List<Int>? = null
            var worryOp: OpPair? = null
            var testOp: OpPair? = null
            var testThrows: ThrowTo? = null

            for (line in lines) {
                if (line.startsWith("Monkey ")) {
                    n = line.substring(7, line.length-1).toInt()
                } else if (line.startsWith("  Starting items: ")) {
                    items = line.substring(18)
                        .split(',')
                        .map { it.trim().toInt() }
                } else if (line == "  Operation: new = old * old") {
                    worryOp = OpPair.pow(2)
                } else if (line.startsWith("  Operation: ")) {
                    val parts = line.substring(13).split(' ')
                    val op = when (parts[3]) {
                        "+" -> Op.Plus
                        "*" -> Op.Multiply
                        else -> Op.Plus
                    }
                    val value = parts.last().toInt()
                    worryOp = OpPair(op, value)
                } else if (line.startsWith("  Test: ")) {
                    val parts = line.substring(8).split(' ')
                    val op = when(parts[0]) {
                        "divisible" -> Op.Divisible
                        else -> Op.Plus
                    }
                    val value = parts.last().toInt()
                    testOp = OpPair(op, value)
                    testThrows = ThrowTo(-1, -1)
                } else if (line.startsWith("    If true: ")) {
                    val ifTrue = line.split(' ').last().toInt()
                    testThrows = testThrows?.copy(ifTrue = ifTrue)
                } else if (line.startsWith("    If false: ")) {
                    val ifFalse = line.split(' ').last().toInt()
                    testThrows = testThrows?.copy(ifFalse = ifFalse)
                }
            }

            return Monkey(
                n,
                items?.toMutableList() ?: mutableListOf(),
                worryOp ?: OpPair.plus(-1),
                testOp ?: OpPair.divisible(1),
                testThrows ?: ThrowTo(-1, -1)
            )
        }
    }
}

data class OpPair(val op: Op, val value: Int) {
    fun exec(current: Int): Int {
        return op.exec(current, value)
    }

    companion object {
        fun plus(value: Int): OpPair {
            return OpPair(Op.Plus, value)
        }

        fun pow(value: Int): OpPair {
            return OpPair(Op.Power, value)
        }

        fun divisible(value: Int): OpPair {
            return OpPair(Op.Divisible, value)
        }
    }
}

enum class Op {
    Plus {
         override fun exec(x: Int, y: Int): Int = x + y
     },
    Subtract {
        override fun exec(x: Int, y: Int): Int = x - y
    },
    Multiply {
        override fun exec(x: Int, y: Int): Int = x * y
    },
    Divide {
        override fun exec(x: Int, y: Int): Int = x / y
    },
    Power {
        override fun exec(x: Int, y: Int): Int = x.toDouble().pow(y).roundToInt()
    },
    Divisible {
        override fun exec(x: Int, y: Int): Int = x % y
    };

    abstract fun exec(x: Int, y: Int): Int
}

data class ThrowTo(val ifTrue: Int, val ifFalse: Int)
data class ThrowInfo(val target: Int, val worryAmount: Int)
